cassandra/policies.py

```diff
@@ -316,6 +327,134 @@ def on_add(self, host):
+
+        if dc == self.local_dc:
+            return HostDistance.LOCAL
+
+        if not self.used_hosts_per_remote_dc:
+            return HostDistance.IGNORED
+        else:
+            dc_hosts = self._dc_live_hosts.get(dc, ())
+            if not dc_hosts:
+                return HostDistance.IGNORED
+
+            if host in list(dc_hosts)[:self.used_hosts_per_remote_dc]:
+
+                return HostDistance.REMOTE
+            else:
+                return HostDistance.IGNORED
+
+    def make_query_plan(self, working_keyspace=None, query=None):
+        pos = self._position
+        self._position += 1
+
+        local_rack_live = self._live_hosts.get((self.local_dc, self.local_rack), ())
+        pos = (pos % len(local_rack_live)) if local_rack_live else 0
+        # Slice the cyclic iterator to start from pos and include the next len(local_live) elements
+        # This ensures we get exactly one full cycle starting from pos
+        for host in islice(cycle(local_rack_live), pos, pos + len(local_rack_live)):
+            yield host
+
+        local_live = tuple(host for host in self._dc_live_hosts.get(self.local_dc, ()) if host.rack != self.local_rack)
+        pos = (pos % len(local_live)) if local_live else 0
+        for host in islice(cycle(local_live), pos, pos + len(local_live)):
+            yield host
+
+        # the dict can change, so get candidate DCs iterating over keys of a copy
+        other_dcs = [dc for dc in self._dc_live_hosts.copy().keys() if dc != self.local_dc]
+        for dc in other_dcs:
+            remote_live = self._dc_live_hosts.get(dc, ())
+            for host in remote_live[:self.used_hosts_per_remote_dc]:
+                yield host
+
+    def on_up(self, host):
+        dc = self._dc(host)
+        rack = self._rack(host)
+        with self._hosts_lock:
+            current_hosts = self._live_hosts.get((dc, rack), ())
+            if host not in current_hosts:
+                self._live_hosts[(dc, rack)] = current_hosts + (host, )
+            current_hosts_dc = self._dc_live_hosts.get(dc, ())
+            if host not in current_hosts_dc:
+                self._dc_live_hosts[dc] = current_hosts_dc + (host, )
+
+    def on_down(self, host):
+        dc = self._dc(host)
+        rack = self._rack(host)
+        with self._hosts_lock:
+            current_hosts = self._live_hosts.get((dc, rack), ())
+            if host in current_hosts:
+                hosts = tuple(h for h in current_hosts if h != host)
+                if hosts:
+                    self._live_hosts[(dc, rack)] = hosts
+                else:
+                    del self._live_hosts[(dc, rack)]
+            current_hosts_dc = self._dc_live_hosts.get(dc, ())
+            if host in current_hosts_dc:
+                hosts = tuple(h for h in current_hosts_dc if h != host)
+                if hosts:
+                    self._dc_live_hosts[dc] = hosts
+                else:
+                    del self._dc_live_hosts[dc]
```

Comment:
Why is there no `del` branch in the previous fragment (handling `self._live_hosts`)?

Reply:
There is one in line 444